chisel/src/dispatcher.rs

```diff
@@ -610,6 +610,71 @@ impl ChiselDisptacher {
                     Err(e) => DispatchResult::CommandFailed(e.to_string()),
                 }
             }
+            ChiselCommand::Edit => {
+                if let Some(session_source) = self.session.session_source.as_mut() {
+                    // create a temp file with the content of the run code
+                    let mut file_path = std::env::temp_dir();
+                    file_path.push("chisel-tmp.sol");
+                    let result = std::fs::File::create(&file_path)
+                        .map(|mut file| file.write_all(session_source.run_code.as_bytes()));
+                    if let Err(e) = result {
+                        return DispatchResult::CommandFailed(format!(
+                            "Could not write to a temporary file: {}",
+                            e
+                        ))
+                    }
+
+                    // open the temp file with the editor
+                    let editor = std::env::var("EDITOR").unwrap_or_else(|_| "vim".to_string());
+                    let mut cmd = Command::new(editor);
+                    cmd.arg(&file_path);
+
+                    match cmd.status() {
+                        Ok(status) => {
+                            if !status.success() {
+                                if let Some(status_code) = status.code() {
+                                    return DispatchResult::CommandFailed(format!(
+                                        "Editor exited with status {}",
+                                        status_code
+                                    ))
+                                } else {
+                                    return DispatchResult::CommandFailed(
+                                        "Editor exited without a status code".to_string(),
+                                    )
+                                }
+                            }
+                        }
+                        Err(_) => {
+                            return DispatchResult::CommandFailed(
+                                "Editor exited without a status code".to_string(),
+                            )
+                        }
+                    }
+
+                    // if the editor exited successfully, try to compile the new code
+                    let edited_code = std::fs::read_to_string(file_path).unwrap();
+                    let mut new_session_source = session_source.clone();
```

Comment:
We may want to expose the `SessionSource`'s `shallow_clone` function for this, but not necessary.